bindings/nodejs/src/sol-uv-integration.c

```diff
@@ -0,0 +1,212 @@
+    .check = uv_loop_source_check,
+    .dispatch = uv_loop_source_dispatch,
+    .dispose = NULL
+};
+
+static bool
+uv_loop_fd_changed(void *data, int fd, uint32_t active_flags)
+{
+    SOL_DBG("Running one uv loop iteration");
+    uv_run(data, UV_RUN_NOWAIT);
+    return true;
+}
+
+int
+hijack_main_loop()
+{
+    int returnValue;
+    uv_loop_t *uv_loop = NULL;
+
+    SOL_DBG("Entering with state %s", RESOLVE_MAINLOOP_STATE(mainloopState));
+    if (mainloopState == MAINLOOP_HIJACKED ||
+        mainloopState == MAINLOOP_HIJACKING_STARTED) {
+        return 0;
+    }
+
+    uv_loop = uv_default_loop();
+
+    // The actual hijacking starts here, inspired by node-gtk. The plan:
+    // 1. Attach a source to the soletta main loop which will run the uv main
+    //    loop in a non-blocking fashion.
+    // 2. Attach an idler to the uv main loop and call sol_run() from it when
+    //    it first runs. This interrupts the uv main loop, because sol_run()
+    //    doesn't return but, since we've already added the above source to the
+    //    soletta main loop in the first step, the source will end up running
+    //    one non-blocking iteration of the uv main loop which, in turn, will
+    //    recursively call the idler we added. At that point, the ilder can
+    //    remove itself from the uv main loop. After that, only the soletta
+    //    main loop runs, but it runs an iteration of the uv main loop in a
+    //    non-blocking fashion whenever the uv main loop signals to the
+    //    soletta main loop via the attached source.
+
+    // We allocate the various needed structures only once. After that, we
+    // reuse them. We never free them, even if we release the uv main loop.
+    if (!uv_loop_source) {
+        uv_loop_source = sol_mainloop_source_add(&uv_loop_source_funcs,
+            uv_loop);
+        if (!uv_loop_source) {
+            return -ENOMEM;
+        }
+    }
+
+    if (!uv_loop_fd) {
+        uv_loop_fd = sol_fd_add(uv_backend_fd(uv_loop),
+            SOL_FD_FLAGS_IN | SOL_FD_FLAGS_OUT | SOL_FD_FLAGS_ERR,
+            uv_loop_fd_changed, uv_loop);
+        if (!uv_loop_fd) {
+            return -ENOMEM;
+        }
+    }
+
+    returnValue = uv_idle_init(uv_loop, &uv_idle);
+    if (returnValue) {
+        return returnValue;
+    }
+
+    SOL_DBG("Starting idler");
+    returnValue = uv_idle_start(&uv_idle, uv_idle_callback);
+    if (returnValue) {
+        return returnValue;
+    }
+
+    mainloopState = MAINLOOP_HIJACKING_STARTED;
+    return 0;
+}
+
+int
+release_main_loop()
+{
+    int returnValue = 0;
+
+    SOL_DBG("Entering with state %s", RESOLVE_MAINLOOP_STATE(mainloopState));
+    if (mainloopState == MAINLOOP_RELEASED ||
+        mainloopState == MAINLOOP_RELEASING_STARTED) {
+        return returnValue;
+    }
+
+    // hijack_main_loop() was called, but the idler has not yet run
+    if (mainloopState == MAINLOOP_HIJACKING_STARTED) {
+        SOL_DBG("idler has not yet run, so stopping it");
```

Comment:
oh yeah i forgot to add the "not"... lol
Sorry...
"Idler has not run yet, so stopping it"